api-authentication/server/index.js

```diff
@@ -0,0 +1,104 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+
+  /* your code starts here */
+  let usernameData, hashedPasswordData, userId;
+  const sql = `
+        select "username",
+               "hashedPassword"
+        from "users"
+        where "hashedPassword" = $2
+        returning "userId", "username", "password"`;
+  const params = [userId, usernameData, hashedPasswordData];
+  if (username !== usernameData) {
+    throw new ClientError(401, 'invalid login');
+  }
+  argon2
+    .verify(hashedPasswordData, password)
+    .then(isMatching => {
+      if (!isMatching) {
+        throw new ClientError(401, 'invalid login');
+      }
+      db.query(sql, params)
+        .then(result => {
+          const payload = { userId, username };
+          const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+          res.status(200).json(token);
+        })
+        .catch(err => next(err));
+    })
+    .catch(err => next(err));
```

Comment:
You need to query the database first to get the `hashedPassword` from the database. Then you use `.verify` to compare the `hashedPassword` to the `password` the user sent to the endpoint.